### PR TITLE
[CMake] Enable CMP0179 alongside CMP0156 for deduplication on LLD

### PR DESCRIPTION
**Pull request by tambry**
LLD has a bug regarding ordering of static link libraries in the ELF backend, which has been reported as #116669.
CMake 3.31.0 started properly deduplicating static libraries for LLD causing the following linking failure for `libclang-cpp.so` with `-DLLVM_LINK_LLVM_DYLIB=ON`:
```
ld.lld: error: undefined symbol: llvm::omp::getOpenMPClauseName(llvm::omp::Clause)
>>> referenced by OpenMPKinds.cpp
>>>               tools/clang/lib/Basic/CMakeFiles/obj.clangBasic.dir/OpenMPKinds.cpp.o:(clang::getOpenMPSimpleClauseTypeName(llvm::omp::Clause, unsigned int))
>>> referenced by SemaOpenMP.cpp
>>>               tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/SemaOpenMP.cpp.o:(clang::SemaOpenMP::CheckOMPRequiresDecl(clang::SourceLocation, llvm::ArrayRef<clang::OMPClause*>))
>>> referenced by SemaOpenMP.cpp
>>>               tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/SemaOpenMP.cpp.o:(clang::SemaOpenMP::CheckOMPRequiresDecl(clang::SourceLocation, llvm::ArrayRef<clang::OMPClause*>))
>>> referenced 166 more times

[tons more]
```

CMake 3.31 also introduced CMP0179, which builds on CMP0156 and makes the deduplication consistent across platforms.
By coincidence this works around the above LLD deficiency and is the fix that CMake 3.31.1 will implement.
However, the fix is to ignore CMP0156 unless CMP0179 is also enabled, i.e. no more deduplication.
So enable CMP0179 to keep the benefits of deduplication from CMP0156 on LLD and fix the build for CMake 3.31.0.

See: #116669
See: https://gitlab.kitware.com/cmake/cmake/-/issues/26447
Fixes: cb90d5b3ef463f0a471f9c6d39978c3764021dea

### CODE REVIEW

File: cmake/Modules/CMakePolicy.cmake
Comment (compnerd) [inline cmake/Modules/CMakePolicy.cmake:48]: I think that this isn't the ideal way to handle this. I think that we should simply limit CMP0156 to `if(APPLE OR WIN32)`. ELFish platforms have the expectation that the command line order and duplication is material and the de-duplication will change the behaviour of the symbolic resolution.

**Review by compnerd — CHANGES_REQUESTED** [review]

File: cmake/Modules/CMakePolicy.cmake
Comment (boomanaiden154) [reply cmake/Modules/CMakePolicy.cmake:48]: These CMake targets will eventually be set to new (unless explicitly set to old) once we bump the minimum required CMake version. It would be nice if when we run across issues we set the policy to make it easier when bumping the CMake version.

Agreed that this is not really the correct solution to the specific problem mentioned here though.

Comment (tambry) [reply cmake/Modules/CMakePolicy.cmake:48]: I don't see a reason to limit this platform-wise. It works fine on all up until CMake 3.31.0 started taking advantage of leeway granted by CMP0156=NEW specification, if I understand correctly. However LLVM is affected by this and requires the more sensible behaviour provided by CMP0179=NEW, which coincides with the above change.

> Agreed that this is not really the correct solution to the specific problem mentioned here though.

What would be the correct solution?  
CMP0179 solves the problem, it'll eventually get set anyway across the board and it changes to behaviour to what LLVM requires to be properly linked.

**Comment by tambry** [comment]
LLD bug has now been reported as https://github.com/llvm/llvm-project/issues/116669 and I've reworded the commit message to hopefully make the effect and intent of this change clearer. Please do have another look.

**Review by boomanaiden154 — APPROVED** [review]
Seems reasonable enough to me. Someone else should approve this too before landing though.

**Comment by tambry** [comment]
Gentle ping @compnerd for another look.

**Review by compnerd — COMMENTED** [review]
I understand that LLD has ignored the requirements for ELF semantics. Please test with BFD. I expect that we need to disable this policy for BFD and gold to successfully link. We can add an additional case for that if you want.

```cmake
if(APPLE OR WIN32 OR LLVM_ENABLE_LLD)
  ...
endif()
```

However, if BFD and gold are able to link properly ignoring the symbol resolution semantics (and the output is the same), then this seems reasonable.

**Comment by etcwilde** [comment]
> I understand that LLD has ignored the requirements for ELF semantics. Please test with BFD. I expect that we need to disable this policy for BFD and gold to successfully link. We can add an additional case for that if you want.
> 
> ```cmake
> if(APPLE OR WIN32 OR LLVM_ENABLE_LLD)
>   ...
> endif()
> ```
> 
> However, if BFD and gold are able to link properly ignoring the symbol resolution semantics (and the output is the same), then this seems reasonable.

CMake added linker detection in 3.29 specifically so that it could handle these cases.
The addition of [CMAKE_LINKER_TYPE](https://cmake.org/cmake/help/v3.29/variable/CMAKE_LINKER_TYPE.html) was part of this. The expectation from CMP0156 is that it only de-dupes the libraries on the link line for linkers that don't require it.
Something regressed in 3.31 for LLD. [CMP0156](https://cmake.org/cmake/help/latest/policy/CMP0156.html) and [CMP0179](https://cmake.org/cmake/help/latest/policy/CMP0179.html) should not affect BFD or gold, as they do require repetition of the libraries.

**Review by etcwilde — APPROVED** [review]
This looks fine to me. Thanks

**Review by compnerd — APPROVED** [review]
I'd missed the "based on linker capabilities" bit for CMP0156. In the case that it is LLD, taking the first occurrence seems fine. The documentation states that this only applies to Apple, Windows, and LLD.

**Comment by tambry** [comment]
Thanks!

> Something regressed in 3.31 for LLD.

Not quite.
CMake wasn't capable of detecting LLD backends before so it didn't apply the deduplication to it. CMake 3.31 added such detection and thus could apply the deduplication. However the LLD backend turned out to be not conformant to its own documentation on this, i.e. it wasn't actually capable of linking when deduplicated as specified. This has been reported as https://github.com/llvm/llvm-project/issues/116669.
CMake 3.31.1 will add a workaround to disable the deduplication for LLD's ELF backend unless CMP0179 is _also_ enabled, which changes the behaviour to be consistent across platforms and happens to work around this LLD bug.

**Comment by MaskRay** [comment]
What is this deduplication issue and do you have a small reproduce? https://github.com/llvm/llvm-project/issues/116669 does not contain a reproduce. Archive semantics are special (https://maskray.me/blog/2021-06-20-symbol-processing#shared-object-overriding-archive)

**Pull request merged**
